homeassistant/components/mqtt/notify.py

```diff
@@ -0,0 +1,397 @@
+
+
+class MqttNotificationConfig(TypedDict, total=False):
+    """Supply service parameters for MqttNotificationService."""
+
+    command_topic: str
+    command_template: Template
+    encoding: str
+    name: str | None
+    qos: int
+    retain: bool
+    targets: list
+    title: str
+    device: ConfigType
+
+
+async def async_initialize(hass: HomeAssistant) -> None:
+    """Initialize globals."""
+    await async_setup_reload_service(hass, DOMAIN, PLATFORMS)
+    hass.data.setdefault(MQTT_NOTIFY_SERVICES_SETUP, {})
+
+
+def _check_notify_service_name(hass: HomeAssistant, config: ConfigType) -> str | None:
+    """Check if the service already exists or else return the service name."""
+    service_name = slugify(config[CONF_NAME])
+    has_services = hass.services.has_service(notify.DOMAIN, service_name)
+    services = hass.data[MQTT_NOTIFY_SERVICES_SETUP]
+    if service_name in services.keys() or has_services:
+        _LOGGER.error(
+            "Notify service '%s' already exists, cannot register service",
+            service_name,
+        )
+        return None
+    return service_name
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up MQTT notify service dynamically through MQTT discovery."""
+    await async_initialize(hass)
+    setup = functools.partial(_async_setup_notify, hass, config_entry=config_entry)
+    await async_setup_entry_helper(hass, notify.DOMAIN, setup, DISCOVERY_SCHEMA)
+
+
+async def _async_setup_notify(
+    hass,
+    legacy_config: ConfigType,
+    config_entry: ConfigEntry,
+    discovery_data: dict[str, Any],
+):
+    """Set up the MQTT notify service with auto discovery."""
+    config = DISCOVERY_SCHEMA(discovery_data[ATTR_DISCOVERY_PAYLOAD])
+    discovery_hash = discovery_data[ATTR_DISCOVERY_HASH]
+
+    if not (service_name := _check_notify_service_name(hass, config)):
+        async_dispatcher_send(hass, MQTT_DISCOVERY_DONE.format(discovery_hash), None)
+        clear_discovery_hash(hass, discovery_hash)
+        return
+
+    device_id = None
+    if CONF_DEVICE in config:
+        await _update_device(hass, config_entry, config)
+        device_registry = await hass.helpers.device_registry.async_get_registry()
+        device = device_registry.async_get_device(
+            {(DOMAIN, id_) for id_ in config[CONF_DEVICE][CONF_IDENTIFIERS]},
+            {tuple(x) for x in config[CONF_DEVICE][CONF_CONNECTIONS]},
+        )
+        device_id = device.id
+
+    service = MqttNotificationService(
+        hass,
+        cast(MqttNotificationConfig, config),
+        config_entry,
+        device_id,
+        discovery_hash,
+    )
+    hass.data[MQTT_NOTIFY_SERVICES_SETUP][service_name] = service
+
+    await service.async_setup(hass, service_name, service_name)
+    await service.async_register_services()
+
+
+def has_notify_services(hass: HomeAssistant, device_id: str) -> bool:
```

Comment:
```suggestion
def _device_has_notify_services(hass: HomeAssistant, device_id: str) -> bool:
```

Move it together with the other helper `_check_notify_service_name` so the code is easier to read

Reply:
Omitted the `_` since the function is called from `mixins.py`